### PR TITLE
docs: added optimization

### PR DESCRIPTION
**Pull request by maximilianwerk**
This is a first try for optimization. 
@florian-hoenicke Please review critically.

### CODE REVIEW

**Comment by maximilianwerk** [comment]
An explanation of `MultiFlowRunner` is missing.

imports are missing

**Review by JoanFM — COMMENTED** [review]
Would it make sense to link the user before to the concept of Evaluators in Jina and how it works with GroundTruths?

I miss also the focus on describing a set of hyperparameters that are the ones to be tuned? maybe it is there but I did not see it too much highlighted?

File: chapters/optimization/index.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+## Flow Optmization
+
+##### Flow Optimization eases the process of rerunning Flows with different parameter sets. It allows out of the box hyperparameter tuning inside Jina.
+
+### Flow Optimization is hyperparameter tuning on steroids
+
+A common principle when building a Flow in Jina is:
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: A common pattern

File: chapters/optimization/index.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+1. Design the high level `flow.yml`
+2. Define several `pods.yml` files for all needed Executors
+3. Repeat until happy with results:
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: Until the evaluation metric suits your use case

File: chapters/optimization/index.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+Flow Optimization automates exactly these steps.
+It can be done via python code or `.yml` definitions as commonly used around Jina.
+
+### Before you start
+
+Make sure you install Jina via `Installation <https://docs.jina.ai/chapters/install/os/index.html>`_ with `jina[optimizer]`.
+
+### Using FlowOptimization
+
+In this toy example, we choose the optimal layer of an encoder for the final embedding.
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: Try to find instead of choose?

File: chapters/optimization/index.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+This is a common practice in ML.
+The best semantic representation for a given problem might not be the last layer of a given model.
+
+Flow Optimization requires the following components:
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: A Flow Opti....

File: chapters/optimization/index.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+- At least one Flow and the corresponding Pod definitions via JAML
+- An Evaluator Executor in at least one of the Flows
+- Documents, which are send to each Flow
+- A `parameter.yml` file describing the optimization scenario
+- A `FlowRunner` object, which allows repeatedly running the same Flow with different configurations.
+
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: A data source, e.g., json lines file

File: chapters/optimization/index.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+!Flow
+version: '1'
+env:
+  JINA_ENCODER_LAYER_VAR: ${{JINA_ENCODER_LAYER}}
+pods:
+  - uses: encoder.yml
+  - uses: EuclideanEvaluator
+```
+
+The `EuclideanEvaluator` is used for calculating the distance between the encoded and an expected vector.
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: the expected


What is the encoded vector?

File: chapters/optimization/index.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+- !IntegerParameter
+  jaml_variable: JINA_ENCODER_LAYER
+  high: 2
+  low: 0
+  step_size: 1
+```
+
+Here we say, the variable can take `int` values in the range `[0, 2]`.
+Jina leverages the optuna optimizer under the hood.
+For a detailed description of all possible choices see (TODO: INSERT LINK TO PARAMETER DOCSTRING).
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: Todo

File: chapters/optimization/index.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    evaluation_callback=MeanEvaluationCallback(),
+    n_trials=3,
+    direction='minimize',
+    seed=1
+)
+result = optimizer.optimize_flow()
+```
+
+The `MeanEvaluationCallback` takes the results of the last Evaluator inside a Flow and averages the results.
+In the above defined Flow it is the solemn `EuclideanEvaluator`.
```
Comment (florian-hoenicke) [inline chapters/optimization/index.md]: Solemn?

**Review by florian-hoenicke — DISMISSED** [review]

**Comment by florian-hoenicke** [comment]
I like the playful way of explaining . Also the example is a nice showcase.
The language at the start is great for a blogpost. But I think for the documentation it is too playful

**Review by Yongxuanzhang — COMMENTED** [review]
For the newly added files I heard that we are encouraged to use rst first instead of Markdown. Or it is ok if we could choose the one most comfortable with?

File: chapters/optimization/index.md (hunk 8 above)
Comment (maximilianwerk) [reply chapters/optimization/index.md]: My english... I thought it means something different :D

**Comment by maximilianwerk** [comment]
> Would it make sense to link the user before to the concept of Evaluators in Jina and how it works with GroundTruths?
> 
> I miss also the focus on describing a set of hyperparameters that are the ones to be tuned? maybe it is there but I did not see it too much highlighted?

Once it is there, linking to Evaluators is a great idea. I add a more detailed description of possible hyperparameters.

> For the newly added files I heard that we are encouraged to use rst first instead of Markdown. Or it is ok if we could choose the one most comfortable with?

@FionnD @lusloher Can you please verify this?

**Comment by FionnD** [comment]
> For the newly added files I heard that we are encouraged to use rst first instead of Markdown. Or it is ok if we could choose the one most comfortable with?

Yes, you can use either. If you are indifferent, please use RST

File: chapters/optimization/index.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
+## Flow Optmization
```
Comment (CatStark) [inline chapters/optimization/index.md]: I think in the new template we need a summary of the topics that will be covered in the doc.

You can add it just like this in rst, not sure how this will be with markdown

.. contents:: Table of Contents
    :depth: 2

File: chapters/optimization/index.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+### Using FlowOptimization
+
+In this toy example, we try to find the optimal layer of an encoder for the final embedding.
+This is a common practice in machine learning.
+The best semantic representation for a given problem might not be the last layer of a given model.
+
+A Flow Optimization requires the following components:
+
+- At least one Flow and the corresponding Pod definitions via JAML
```
Comment (CatStark) [inline chapters/optimization/index.md]: Should we link important terms (Flow, Pod, etc) the first time they are mentioned to the Glossary?

File: chapters/optimization/index.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+- An Evaluator Executor in at least one of the Flows
+- A datasource containting Documents, which are send to each Flow
```
Comment (CatStark) [inline chapters/optimization/index.md]: Should be "sent" instead of "send"

File: chapters/optimization/index.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+```python
+documents = [
+    (Document(content='🐲'), Document(embedding=np.array([2]))),
+    (Document(content='🐦'), Document(embedding=np.array([3]))),
+    (Document(content='🐢'), Document(embedding=np.array([3])))
+]
+```
+
+Documents will be send in pairs `(doc, groundtruth)` to the Flow.
```
Comment (CatStark) [inline chapters/optimization/index.md]: "will be sent"

**Review by JoanFM — COMMENTED** [review]
I am missing some reference to HyperParameterOptimization as a `field`?

File: chapters/optimization/index.md (hunk 10 above)
Comment (maximilianwerk) [reply chapters/optimization/index.md]: I won't do it yet. If you think about Flow Optimization, they should be familiar enough already. Anyhow, if we decide to do it in general, I'll add it for sure.

**Comment by FionnD** [comment]
Hi @maximilianwerk what's the latest status on this? Is it ready to be merged or still work to be done?

**Comment by maximilianwerk** [comment]
I will clean up all the annotation (like `EuclideanEvaluator` => :class:`EuclideanEvaluator`) in the evening and ping you then.

File: chapters/optimization/index.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        'pokemon': [2, 4, 7],
+        'dog': [0, 2, 5],
+    }
+
+    def __init__(self, layer=0, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._layer = layer
+
+    def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
+        return np.array([[self.ENCODE_LOOKUP[data[0]][self._layer]]])
```
Comment (davidbp) [inline chapters/optimization/index.md]: `self.ENCODE_LOOKUP` is a dict where the keys are strings and `data` is a `np.ndarray`.

It seems a bit odd that `data[0]` is used as a key for `ENCODE_LOOKUP ` because  `data` is a `np.ndarray`.

 It is true that `np.ndarray` migtht have as `dtype` a string but that is a very uncommon use. If that is the case I might write it explicitly.

If the input is not necessarily a `np.ndarray` object, it could be written as `Iterable[str]` which would consider the case of `np.ndarray` of  strings.
```
def encode(self, data:`Iterable[str]`, *args, **kwargs) -> 'np.ndarray':
         return np.array([[self.ENCODE_LOOKUP[data[0]][self._layer]]])
```

File: chapters/optimization/index.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+    ENCODE_LOOKUP = {
+        'cat': [1, 3, 5],
+        'pokemon': [2, 4, 7],
+        'dog': [0, 2, 5],
+    }
+
+    def __init__(self, layer=0, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._layer = layer
```
Comment (davidbp) [inline chapters/optimization/index.md]: `._layer` refers to the column in the embedding of each word in `ENCODE_LOOKUP`. Since normally `layer` refers to a particular layer of a neural net, what do you think about calling this integer a name such as `col` or `column` ?  What this variable is doing is selecting a column from the embedding vector.

**Review by davidbp — DISMISSED** [review]
The explanation is pretty col, I would simply change  
- I would  `layer`  for something like `col` 
- I would not write as typehint a `np.ndarray`  with dtype `str`.

File: chapters/optimization/index.md (hunk 13 above)
Comment (maximilianwerk) [reply chapters/optimization/index.md]: Great point. Apparently `data: Sequence[str]`, otherwise we couldn't check and index.

File: chapters/optimization/index.md (hunk 14 above)
Comment (maximilianwerk) [reply chapters/optimization/index.md]: Yes true. Furthermore, I'll add in the text, that this column represents a layer in the model.

**Review by FionnD — APPROVED** [review]
